File: Golang/447. Number of Boomerangs.go
func numberOfBoomerangs(points [][]int) int {
    total := 0
    for _, point1 := range points {
        distanceDict := make(map[int]int, 500)
        for _, point2 := range points {
            distanceDict[distance(point1, point2)]++
        }
        for d := range distanceDict {
            total += distanceDict[d] * (distanceDict[d] - 1)
            total -= distanceDict[0] * (distanceDict[0] - 1)
        }
    }
    return total
}

func distance(a, b []int) int {
    x := a[0] - b[0]
    y := a[1] - b[1]
    return x*x + y*y
}

File: Python3/447. Number of Boomerangs.py
class Solution:
    def numberOfBoomerangs(self, points):
        """
        :type points: List[List[int]]
        :rtype: int
        """
        from functools import partial # 1428ms
        def distance(a, b):
            return (a[0] - b[0])**2 + (a[1] - b[1])**2
        total = 0
        for point in points:
            getDistance = partial(distance, point)
            distances = map(getDistance, points)
            distanceDict = collections.Counter(distances)
            selfcount = distanceDict[0]
            total -= selfcount * (selfcount - 1)
            for count in distanceDict.values():
                total += count * (count - 1)
        return total
